packages/eslint-plugin/src/rules/dependency-checks.ts

```diff
@@ -0,0 +1,325 @@
+
+    if (!projectGraph) {
+      return {};
+    }
+
+    const sourceProject = findProject(
+      projectGraph,
+      projectRootMappings,
+      sourceFilePath
+    );
+
+    // check if source project exists, it's library and has a build target
+    if (!sourceProject || sourceProject.type !== 'lib') {
+      return {};
+    }
+
+    // check if library has a build target
+    const buildTarget = buildTargets.find(
+      (t) => sourceProject.data.targets?.[t]
+    );
+    if (!buildTarget) {
+      return {};
+    }
+
+    // gather helper dependencies for @nx/js executors
+    const helperDependencies = getHelperDependenciesFromProjectGraph(
+      workspaceRoot,
+      sourceProject.name,
+      projectGraph
+    );
+
+    // find all dependencies for the project
+    const npmDeps = findProjectsNpmDependencies(
+      sourceProject,
+      projectGraph,
+      buildTarget,
+      {
+        helperDependencies: helperDependencies.map((dep) => dep.target),
+      },
+      removePackageJsonFromFileMap(projectFileMap)
+    );
+    const projDependencies = {
+      ...npmDeps.dependencies,
+      ...npmDeps.peerDependencies,
+    };
+    const projDependencyNames = Object.keys(projDependencies);
+
+    const projPackageJsonPath = join(
+      workspaceRoot,
+      sourceProject.data.root,
+      'package.json'
+    );
+    let projPackageJsonDeps: Record<string, string> = {};
+    if (existsSync(projPackageJsonPath)) {
+      projPackageJsonDeps = getAllDependencies(projPackageJsonPath);
+    }
+
+    const rootPackageJsonDeps = getAllDependencies(
+      join(workspaceRoot, 'package.json')
+    );
+
+    function validateMissingDependencies(node: AST.JSONProperty) {
+      if (checkMissingDependencies) {
+        const missingDeps = projDependencyNames.filter(
+          (d) => !projPackageJsonDeps[d]
+        );
+
+        missingDeps.forEach((d) => {
+          if (!ignoredDependencies.includes(d)) {
+            context.report({
+              node: node as any,
+              messageId: 'missingDependency',
+              data: { packageName: d },
+              fix(fixer) {
+                const deps = (node.value as AST.JSONObjectExpression)
+                  .properties;
+                if (deps.length) {
+                  return fixer.insertTextAfter(
+                    deps[deps.length - 1] as any,
+                    `,\n    "${d}": "${
+                      rootPackageJsonDeps[d] || projDependencies[d]
+                    }"`
+                  );
+                } else {
+                  return fixer.replaceTextRange(
+                    [node.value.range[0] + 1, node.value.range[1] - 1],
+                    `\n    "${d}": "${
+                      rootPackageJsonDeps[d] || projDependencies[d]
+                    }"\n  `
+                  );
+                }
+              },
+            });
+          }
+        });
+      }
+    }
+
+    function validateInvalidDependencies(node: AST.JSONProperty) {
```

Comment:
Should this be broken into separate `validateObseleteDependencies` and `validateMismatchedDependencies`? An the if check can be a guard to remove a level of nesting. Separating the functions is more consistent since the other validate functions are already separate. Either combine them all, or separate them all IMO.

Reply:
They are currently split based on the node path. It would still have to be single `validateInvalidDependencies` function but can be then split into two separate sub-functions